src/coreclr/tools/Common/TypeSystem/CodeGen/IJitHashableOnly.cs

```diff
@@ -0,0 +1,41 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using Internal.TypeSystem;
+
+namespace Internal.TypeSystem
+{
+    public class JitObjectComparer : IEqualityComparer<object>
+    {
+        public new bool Equals(object x, object y) => x == y;
+        public int GetHashCode(object obj)
+        {
+            if (obj is IJitHashableOnly jitHashable)
+                return jitHashable.GetJitVisibleHashCode();
+            return obj.GetHashCode();
+        }
+    }
+
+    public class JitObjectComparer<T> : IEqualityComparer<T> where T:class
+    {
+        public bool Equals(T x, T y) => x == y;
+        public int GetHashCode(T obj)
+        {
+            if (obj is IJitHashableOnly jitHashable)
+                return jitHashable.GetJitVisibleHashCode();
+            return obj.GetHashCode();
+        }
+    }
+
+    // Mark a type system object with this interface to indicate that it
```

Comment:
I'm not sure I understand how this avoids putting the objects into long lived storage - we call a different GetHashCode - how does that avert long lived storage?

Reply:
Ah, the trick is to make the normal GetHashCode fail with an exception so that special logic needs to be written to make it be stuffable into a collection, but allow a separate path for the hashcode for situations where the JitInterface needs to store the value into a hashtable

Reply:
Ah, so this is just a way to guard these instances don't leak out of CorInfoImpl and don't get stored in one of the many collections we have.

Should IJistHashtableOnly go to src\coreclr\tools\Common\JitInterface (and the respective namespace)? It looks odd to include this from the type system project if the purpose is specifically not to make this a general type system entity.